### ISSUE TITLE
Failure when using activities with biosphere exchanges

### ISSUE THREAD

**Issue by simaosr** [oie-mines-paristech/lca_algebraic#39]
When working with an LCA model on which there are exchanges directly with the biosphere, lca_algebraic fails due to the "formula" field, that means something else.

Changing, in, **_getAmountOrFormula**, the line

`    if "formula" in ex :`
for
`    if "formula" in ex and ex["type"] != "biosphere":`

solves it for me.

Is there another way of handling activities like this, or this is one right solution?

Thank you.
Simão Rodrigues

**Comment by raphaeljolivet** [comment]
Hi, 

Normally, when stumbling upon biosphere flux, the library creates a dummy activity in the technosphere with a single biosphere flux, in order to be able to ask LCIA impact for it to brightway.

This is done in the method [_createTechProxyForBio](https://github.com/oie-mines-paristech/lca_algebraic/blob/3694ca757b7fed119dc51732e30a831ae792f1ac/lca_algebraic/lca.py#L489) that checks the db name is "biosphere3" or that the its type is either "emission" or "natural resource".

Also, please ensure that only a single database is marked as foreground, and that neither ecoinvent or the biosphere are marked as foreground db. See [setForeground](https://oie-mines-paristech.github.io/lca_algebraic/doc/helpers.html#lca_algebraic.helpers.setForeground) and [setBackground](https://oie-mines-paristech.github.io/lca_algebraic/doc/helpers.html#lca_algebraic.helpers.setBackground)

**Comment by simaosr** [comment]
Hi,

Thank you for answering.

 I am working with an LCA that has some activities that are adapted from ecoinvent, and in the biosphere flows have the "formula" field filled with its chemical formula. They are in the only foreground DB in this project.

I have the code failing in _actToExpression_ inside __modelToExpr_, so before __createTechProxyForBio_ is called.  The expression that is being built is based on the chemical formulas, which are seen as parameters.

I see that in bw2.5, for elementary flows, the "formula" field is correctly assigned to the mathematical relationship, and there is another field "chemical formula", but not in bw2.

Best regards,
Simão

**Comment by raphaeljolivet** [comment]
This kind of error is usually catched and ignored : 
It simply writes a warning message that you could ignore as well.

See :

```
def _getAmountOrFormula(ex: ExchangeDataset) -> Union[Basic, float]:
    """ Return either a fixed float value or an expression for the amount of this exchange"""
    if 'formula' in ex:
        try:
            return _parse_formula(ex['formula'])
        except Exception as e:
            error("Error while parsing formula '%s' : backing to amount" % ex['formula'])
```

**Comment by simaosr** [comment]
Yes, I get these warnings, and then it fails in:

```python
            res += formula * act_expr * avoidedBurden
```
with
`TypeError: unsupported operand type(s) for *: 'SingletonRegistry' and 'Symbol'` 

at this point, _formula_ is "S".

_parse_formula is able to parse a few of the chemical formulas as mathematical formulas and is creating the expression using those.

**Comment by raphaeljolivet** [comment]
Ok, I never encountered this case. Biosphere formulate were catched by this mecanism before.

Then I guess you fix is a good one !
Does it work after that ?

**Comment by simaosr** [comment]
Yes, with that fix everything works fine for me.

**Comment by raphaeljolivet** [comment]
Then thanks for this feedback.
I will integrate it in a next release that I plan to do at the beginning of 2024.

Have a good day.

**Comment by raphaeljolivet** [comment]
For the record, I won't include this fix in the main branch, since it make impossible to parametrize biosphere flux with formulas.
It seems Brightway fixed in with the package [bw2io 0.9.DEV7](https://github.com/brightway-lca/brightway2-io/blob/main/CHANGES.md#09dev7-2022-01-11), by using *chemical-formula* instead of *formula*.
